crates/randcast-mock-demo/src/main.rs

```diff
@@ -0,0 +1,239 @@
+use dkg_core::primitives::{
+    group::{Group, Node},
+    joint_feldman,
+    types::DKGOutput,
+};
+use dkg_core::{DKGPhase, Phase2Result};
+use rand::rngs::ThreadRng;
+use randcast_mock_demo::contract::*;
+use randcast_mock_demo::test_helpers::InMemoryBoard;
+use std::collections::HashMap;
+use threshold_bls::{
+    curve::bls12381::{self, PairingCurve as BLS12_381},
+    group::Curve,
+    poly::Idx,
+    sig::{G1Scheme, Scheme, SignatureScheme, ThresholdScheme},
+};
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    let initial_entropy = 0x8762_4875_6548_6346;
+
+    println!(
+        "controller is deploying... initial entropy: {}",
+        initial_entropy
+    );
+
+    let mut controller = Controller::new(initial_entropy);
+
+    let (t, n) = (3, 5);
+
+    println!("nodes setup... t: {} n: {}", t, n);
+
+    let rng = &mut rand::thread_rng();
+
+    let (mut board, phase0s) = setup::<bls12381::Curve, G1Scheme<BLS12_381>, _>(n, t, rng);
+
+    println!("nodes are registering to controller...");
+
+    phase0s.iter().enumerate().for_each(|(i, phase0)| {
+        controller.node_register(
+            String::from("0x") + &i.to_string(),
+            bincode::serialize(&phase0.info.public_key).unwrap(),
+            String::from(""),
+            String::from("0x") + &i.to_string(),
+        );
+    });
+
+    println!("DKG task is emitting...");
+
+    let dkg_task = controller.emit_dkg_task();
+
+    let group_index = dkg_task.group_index;
+
+    let group_epoch = dkg_task.epoch;
+
+    // executes the DKG state machine and ensures that the keys are generated correctly
+    let outputs = run_dkg::<bls12381::Curve, G1Scheme<BLS12_381>>(&mut board, phase0s, rng).await;
+
+    // get the public key (we have already checked that all outputs' pubkeys are the same)
+    let public_poly = outputs[0].public.clone();
+
+    let pubkey = public_poly.public_key();
+
+    println!("DKG result is committing...");
+
+    (0..n).for_each(|i| {
+        let res = controller.commit_dkg(
+            String::from("0x") + &i.to_string(),
+            group_index,
+            group_epoch,
+            bincode::serialize(&pubkey).unwrap(),
+            bincode::serialize(&public_poly.eval(i as u32).value).unwrap(),
+            vec![],
+        );
+        println!("{}-res: {}", i, res);
+    });
+
+    let group = controller.get_group(1);
+
+    println!("group state: {}", group.state);
+
+    println!("group committers: {:?}", group.committers);
+
+    // let msg = rand::random::<[u8; 32]>().to_vec();
+
+    let msg = String::from("ujehwsndfgljkhrlkg");
+
+    println!("An user is requesting a randomness... msg seed: {}", msg);
+
+    let request_res = controller.request(msg.clone());
```

Comment:
maybe `request`'s parameter can be change to `&str`, or if the msg is no longer needed, move the ownership into the request is ok.